drivers/flash/flash_npcx_fiu_nor.c

```diff
@@ -0,0 +1,551 @@
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void flash_npcx_nor_pages_layout(const struct device *dev,
+					const struct flash_pages_layout **layout,
+					size_t *layout_size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+
+	*layout = &config->layout;
+	*layout_size = 1;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+static int flash_npcx_nor_read(const struct device *dev, off_t addr,
+				 void *data, size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	struct flash_npcx_nor_data *dev_data = dev->data;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || ((size + addr) > config->flash_size)) {
+		return -EINVAL;
+	}
+
+	/* Lock/Unlock SPI bus also for DRA mode */
+	qspi_npcx_fiu_mutex_lock_configure(config->qspi_bus, &config->qspi_cfg,
+					   dev_data->operation);
+
+	/* Trigger Direct Read Access (DRA) via reading memory mapped-address */
+	memcpy(data, (void *)(config->mapped_addr + addr), size);
+
+	qspi_npcx_fiu_mutex_unlock(config->qspi_bus);
+
+	return 0;
+}
+
+static int flash_npcx_nor_erase(const struct device *dev, off_t addr,
+				  size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	int ret = 0;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || ((size + addr) > config->flash_size)) {
+		return -EINVAL;
+	}
+
+	/* address must be sector-aligned */
+	if (!SPI_NOR_IS_SECTOR_ALIGNED(addr)) {
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sectors */
+	if ((size % config->min_erase_size) != 0) {
+		return -EINVAL;
+	}
+
+	while (size > 0) {
+		flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_WREN);
+
+		if (size == config->flash_size) {
+			/* chip erase */
+			flash_npcx_uma_cmd_only(dev, SPI_NOR_CMD_CE);
+			size -= config->flash_size;
+		} else {
+			uint8_t opcode;
+
+			if (config->min_erase_size == KB(4)) {
+				opcode = SPI_NOR_CMD_SE;
+			} else if (config->min_erase_size == KB(64)) {
+				opcode = SPI_NOR_CMD_BE;
+			} else {
+				return -EINVAL;
+			}
+
+			flash_npcx_uma_cmd_by_addr(dev, opcode, addr);
+
+			addr += config->min_erase_size;
+			size -= config->min_erase_size;
+		}
+
+		ret = flash_npcx_nor_wait_until_ready(dev);
+		if (ret != 0) {
+			break;
+		}
+	}
+
+	return ret;
+}
+
+static int flash_npcx_nor_write(const struct device *dev, off_t addr,
+				  const void *data, size_t size)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	uint8_t *tx_buf = (uint8_t *)data;
+	int ret = 0;
+
+	/* Out of the region of nor flash device? */
+	if ((addr < 0) || ((size + addr) > config->flash_size)) {
+		return -EINVAL;
+	}
+
+	while (size > 0) {
+		size_t sz_write = size;
+
+		/* Don't write more than a page. */
+		if (sz_write >= SPI_NOR_PAGE_SIZE) {
+			sz_write = SPI_NOR_PAGE_SIZE;
+		}
+
+		/* Don't write across a page boundary */
+		if (((addr + sz_write - 1U) / SPI_NOR_PAGE_SIZE)
+							!= (addr / SPI_NOR_PAGE_SIZE)) {
+			sz_write = SPI_NOR_PAGE_SIZE - (addr % SPI_NOR_PAGE_SIZE);
+		}
```

Comment:
All this code is only meaningful for the first loop iteration, which basically means that it can be executed prior to entering the loop.